enhancements/kube-apiserver/generic-delete-protection.md

```diff
@@ -0,0 +1,233 @@
+---
+title: generic-delete-protection
+authors:
+  - "@deads2k"
+reviewers:
+  - "@sttts"
+approvers:
+  - "@derekwaynecarr"
+creation-date: 2020-06-17
+last-updated: 2020-06-17
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - https://github.com/open-cluster-management/backlog/issues/2648#issuecomment-645170280  
+  - https://github.com/open-cluster-management/backlog/issues/2348#issuecomment-642231925  
+replaces:
+superseded-by:
+---
+
+# Generic Delete Protection
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions 
+
+1. At this moment, adoption is racy because deployment.apps/name is insufficient to distinguish based on UID, but to allow
+   freedom of creation ordering, the CriticalService cannot include a UID.
+   We may be able to create a status field that tracks a UID, but the intent isn't fully clear.
+   If adoption doesn't work, we can end up in a state where ManifestWork cannot be finalized because the CriticalService
+   isn't actually eligible for deletion.
+   The simplest solution seem to be to state an intent of "best-effort-deletion" or "orphan-resources".
+   This could be added as a spec field which could be set post-deletion, pre-finalization.
+
+## Summary
+
+By running the platform on the cluster, there are complications that arise regarding deletion of certain critical services.
+These services include webhook admission plugins and finalizers.
+This design is about trying to find a way to prevent permanently "stuck" resources without trying to coordinate resource deletion
+in clients.
+
+## Motivation
+
+If pods for admission-webhook-a, which protect resource-a, are deleted, then it becomes impossible to create, update, or
+delete any resource-a until the webhookconfiguration is deleted or the pods are restored.
+If pods for finalizer-a, which protect resource-a, are deleted, then it becomes impossible to delete any resource-a until
+the pods are restored.
+Resolving resource deletion ordering in clients is impractical.
+
+### Goals
+
+1. Eliminate the need for ordered client deletion.
+
+### Non-Goals
+
+1. Prevent permanently stuck resources in cases where the resource that is stuck and the the pod providing the critical service
+   are in the same namespace.  Don't do that. 
+
+## Proposal
+
+We will create a validating admission webhook that intercepts DELETEs of namespaces, criticalservices, and any resource listed as a .spec.provider.
```

Comment:
which means it mutates its own admission registration object depending on which provider resources are defined?